[READY] Adds Phoron Sublimation

CakeQ opened the pull request:
🆑 Cakey
rscadd: Adds phoron sublimation, the process of combining shards of supermatter together to form purer concentrations of crystals using gaseous phoron.
rscadd: Adds the resonance chamber, a machine capable of suspending reactions within in order to safely store supermatter cores.
rascadd: Adds the neutron furnace, a machine used to convert quantities of supermatter into other forms of materials with effort. The future of alchemy is here.
rscadd: Adds phoroweave gloves, used to safely touch supermatter without getting vaporised.
tweak: Supermatter cores now have seven tiers, with their characteristics increasing with each tier, giving better power output at the risk of higher instability.
maptweak: Added the sublimation lab to the Petrov.
/🆑

At the moment metaphoron (the substance used to form higher tiers of supermatter cores) can only be obtained from smashing the crystals formed from supermatter explosions.
I'd recommend we mapped/edited some away sites to have these crystals in order to entice science players to go on away sites to collect metaphoron.

Preview:
![image](https://user-images.githubusercontent.com/17550641/46377373-5052e400-c690-11e8-986f-78618db0f72d.png)

![image](https://user-images.githubusercontent.com/17550641/46377383-5943b580-c690-11e8-9ba8-79d507375ce5.png)

![image](https://user-images.githubusercontent.com/17550641/46377617-f6065300-c690-11e8-937e-241ffd7f0422.png)

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,135 @@
+	for( var/sm_piece in sm_pieces )
+		qdel( sm_piece )
+
+	..()
```
Need to return the result of `/Destroy()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,135 @@
+				user.visible_message("[user] starts unfastening the bolts on \the [src].", "You start unfastening the bolts on \the [src].")
+
+		if(do_after(user, 20))
+			if(!src) return
```
`if(QDELETED(src)) return`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,135 @@
+		else
+			user << "<span class='notice'>The machine is full!</span>"
+	else
+		user << "<span class='notice'>This machine only accepts supermatter.</span>"
```
`to_chat()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,135 @@
+			sm_pieces += B
+			user << "You put [B] into the machine."
+		else
+			user << "<span class='notice'>The machine is full!</span>"
```
`to_chat()`

MistakeNot4892 replied:
`to_chat(user, SPAN_WARNING(blah))`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,135 @@
+			user.drop_item()
+			B.loc = src
+			sm_pieces += B
+			user << "You put [B] into the machine."
```
`to_chat()`

MistakeNot4892 replied:
`to_chat(user, SPAN_NOTICE(blah))`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,135 @@
+
+	if(istype( B, /obj/item/weapon/shard/metaphoron ))
+		if( sm_pieces.len < max_pieces )
+			user.drop_item()
```
```
if(!user.drop_item()) 
    return
```

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,135 @@
+	if(istype( B, /obj/item/weapon/shard/metaphoron ))
+		if( sm_pieces.len < max_pieces )
+			user.drop_item()
+			B.loc = src
```
`B.forceMove(src)`

MistakeNot4892 replied:
`B.forceMove(src)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,135 @@
+		if( istype( AM, /obj/machinery/power/supermatter ))
+			if( sm_pieces.len < max_pieces )
+				var/obj/machinery/power/supermatter/SM = AM
+				SM.loc = src
```
`SM.forceMove(src)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,135 @@
+	var/list/sm_pieces = list()
+
+/obj/machinery/phoron_desublimer/resonant_chamber/Destroy()
+	for( var/sm_piece in sm_pieces )
```
Need to also remove the entries from the list itself.
Could use the handy `QDEL_NULL_LIST()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,135 @@
+	if( !anchored )
+		return
+
+	var/choice = input( user, "What would you like to do?", \
```
Need to check that the user is still able to continue running `attack_hand()` after `input()`

PsiOmegaDelta replied:
Also need to re-check all pre-conditions, such as `anchored` and powered

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,135 @@
+			if( !active )
+				combine()
+			else
+				user << "<span class='notice'>This machine is already active!</span>"
```
`to_chat()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,135 @@
+	if( !sm_piece )
+		return
+
+	sm_piece.loc = get_turf( src )
```
`sm_piece.dropInto(loc)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,160 @@
+/obj/machinery/computer/phoron_desublimer_control/New()
+	..()
+
+	spawn( 2 )
```
You may not use `spawn()` or `sleep()` in `/New()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,160 @@
+	var/obj/machinery/phoron_desublimer/vessel/vessel
+	var/obj/machinery/phoron_desublimer/furnace/furnace
+
+/obj/machinery/computer/phoron_desublimer_control/New()
```
`/Initialize()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,160 @@
+	var/obj/machinery/phoron_desublimer/furnace/furnace
+
+/obj/machinery/computer/phoron_desublimer_control/New()
+	..()
```
`. = ..()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,160 @@
+	var/active = 0
+	var/assembled = 0
+	var/state = "vessel"
+	var/list/mat_presets = list(   "Steel" = 30,
```
Replace with material name defines.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,160 @@
+
+/obj/machinery/computer/phoron_desublimer_control/Topic(href, href_list)
+	if(..())
+		return
```
`return 1`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,160 @@
+			furnace.produce()
+
+
+	nanomanager.update_uis(src)
```
No. 
`return 1` if the UI needs to be updated.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,160 @@
+
+
+	nanomanager.update_uis(src)
+	add_fingerprint(usr)
```
`..()` in `/Topic()` already handles this.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,160 @@
+									"Osmium" = null )
+
+	var/obj/machinery/phoron_desublimer/vessel/vessel
+	var/obj/machinery/phoron_desublimer/furnace/furnace
```
These parts need to be handled in `/Destroy()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,160 @@
+		if( istype( PD, /obj/machinery/phoron_desublimer/furnace ))
+			furnace = PD
+
+	return
```
Unnecessary return

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,160 @@
+	return 1
+
+/obj/machinery/computer/phoron_desublimer_control/proc/set_preset()
+	if( !usr ) return
```
Whoever calls `set_preset()` should instead be providing a `user`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,160 @@
+	if( !usr ) return
+
+	var/preset = null
+	preset = input( usr, "Which preset would you like set?", "Select Preset", preset ) in mat_presets
```
Need to check that the `user` (not `usr)` is still conscious, able and adjacent after `input()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,160 @@
+	if(..())
+		return
+
+	if(stat & (NOPOWER|BROKEN))
```
Move this check to `/obj/machinery/computer/phoron_desublimer_control/CanUseTopic()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+
+	var/list/mat = list( "Osmium", "Phoron", "Diamonds", "Platinum", "Gold", "Uranium",  "Silver", "Steel", "Supermatter" )
+	var/list/mat_mod = list(    "Steel" = 3.5,
+								"Silver" = 1.5,
```
Replace with material name defines

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+													"Phoron" = /obj/item/stack/sheet/mineral/phoron,
+													"Supermatter" = /obj/item/weapon/shard/supermatter ) // cost per each mod # of bars
+
+/obj/machinery/phoron_desublimer/furnace/New()
```
`/Initialize()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+													"Supermatter" = /obj/item/weapon/shard/supermatter ) // cost per each mod # of bars
+
+/obj/machinery/phoron_desublimer/furnace/New()
+		..()
```
`. = ..()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+	if( !shard )
+		return
+
+	shard.loc = get_turf( src )
```
`shard.DropInto(loc)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+	var/min_neutron_flow = 1
+	var/neutron_flow = 30
+	var/max_neutron_flow = 350
+	var/obj/item/weapon/shard/supermatter/shard = null
```
Needs to be handled in `/Destroy()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+	shard = null
+	update_icon()
+
+/obj/machinery/phoron_desublimer/furnace/proc/modify_flow(var/change)
```
This method is never appears to be called.
Unnecessary code.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+		return
+	if(istype(B, /obj/item/weapon/shard/supermatter))
+		if( !shard )
+			user.drop_item()
```
``` 
if(!user.drop_item())
    return
```

afterthought2 replied:
`unequip_item()`, actually. Can also do
```
if(!user.unequip_item(src)) return
```
and kill the forceMove on the next line.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+	if(istype(B, /obj/item/weapon/shard/supermatter))
+		if( !shard )
+			user.drop_item()
+			B.loc = src
```
`B.forceMove(src)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+			user.drop_item()
+			B.loc = src
+			shard = B
+			user << "You put [B] into the machine."
```
`to_chat()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+			shard = B
+			user << "You put [B] into the machine."
+		else
+			user << "There is already a shard in the machine."
```
`to_chat()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+		else
+			user << "There is already a shard in the machine."
+	else
+		user << "<span class='notice'>This machine only accepts supermatter shards</span>"
```
`to_chat()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,119 @@
+	if( !loaded_tank )
+		return
+
+	loaded_tank.loc = get_turf( src )
```
`loaded_tank.dropInto(loc)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,119 @@
+	if( !loaded_shard )
+		return
+
+	loaded_shard.loc = get_turf( src )
```
`loaded_shard.dropInto(loc)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,119 @@
+	if( icon_state != "ProcessorFull" )
+		flick("ProcessorFill", src)
+		sleep(12)
+		icon_state = "ProcessorFull"
```
Don't change `icon_state` outside of `/New()`, `/Initialize()` or `/update_icon()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,119 @@
+		else
+			user << "There is already a tank in the machine."
+	else
+		user << "/red That's not a valid item!"
```
\red is a forbidden macro (also note how the / above is in the wrong direction)

PsiOmegaDelta replied:
`to_chat()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,119 @@
+			loaded_tank = B
+			user << "You put [B] into the machine."
+		else
+			user << "There is already a tank in the machine."
```
`to_chat()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,119 @@
+			user.drop_item()
+			B.loc = src
+			loaded_tank = B
+			user << "You put [B] into the machine."
```
`to_chat()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,119 @@
+	else if( istype( B, /obj/item/weapon/tank ))
+		if( !loaded_tank )
+			user.drop_item()
+			B.loc = src
```
`B.dropInto(loc)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,119 @@
+			user << "There is already a shard in the machine."
+	else if( istype( B, /obj/item/weapon/tank ))
+		if( !loaded_tank )
+			user.drop_item()
```
```
if(!user.drop_item())
    return
```

afterthought2 replied:
See `unequip` above.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,119 @@
+		return
+	else if(istype(B, /obj/item/weapon/shard/supermatter))
+		if( !loaded_shard )
+			user.drop_item()
```
```
if(!user.drop_item())
    return
```

afterthought2 replied:
See `unequip` above.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,119 @@
+	else if(istype(B, /obj/item/weapon/shard/supermatter))
+		if( !loaded_shard )
+			user.drop_item()
+			B.loc = src
```
`B.dropInto(loc)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,119 @@
+	desc = "Grows supermatter shards by seeding them with phoron."
+	icon_state = "ProcessorEmpty"
+	var/obj/item/weapon/tank/loaded_tank
+	var/obj/item/weapon/shard/supermatter/loaded_shard
```
These objects need to be handled in `/Destroy()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,119 @@
+	icon_state = "ProcessorEmpty"
+	var/obj/item/weapon/tank/loaded_tank
+	var/obj/item/weapon/shard/supermatter/loaded_shard
+	var/datum/gas_mixture/air_contents
```
As does this datum

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,119 @@
+
+	active_power_usage = 10000
+
+/obj/machinery/phoron_desublimer/vessel/New()
```
`/Initialize()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,119 @@
+	active_power_usage = 10000
+
+/obj/machinery/phoron_desublimer/vessel/New()
+	..()
```
`. = ..()`

Kwask commented:
my code lives again 🤖

MistakeNot4892 commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,135 @@
+#define SM_CORE_SIZE 100
+
+/obj/machinery/phoron_desublimer/resonant_chamber
+	name = "Resonant Chamber"
```
Should be lowercase.

MistakeNot4892 commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,135 @@
+
+/obj/machinery/phoron_desublimer/resonant_chamber
+	name = "Resonant Chamber"
+	desc = "Uses subspace resonance to restructure multiple pieces of supermatter into one. It freezes time within, and can be unbolted from the floor, so it is also an ideal container to transport dangerous supermatter."
```
This should probably be in the extended desc rather than the atom desc.

CakeQ committed:
Adds Phoron Desublimation

Progress

more changes

tweaks

final machinery tweaks

merges shard icons

tweaks glove desc

Adds sublimation lab

fixes indentation, updates test

updates

updates resonant chamber description

remove stray var

update checksum

Fixes

updates << count

fixes global test

fixes

fixes

replaces qdel_self

fixes var name

tweaks

fuxes

removes update_icon overrides

replaces update_icons

replaces update_icons

gets rid of useless ..()s

Addresses comments

code cleanup

removes whitespace error

CakeQ commented:
build failing due to something unrelated to PR.
Devs plz

afterthought2 commented on code/modules/supermatter/supermatter_tiers.dm:
```diff
@@ -0,0 +1,176 @@
+
+*/
+
+var/global/list/supermatter_tiers = list(	new /datum/sm_control/level_1,
```
Use `GLOBAL_LIST_INIT` instead.

PsiOmegaDelta commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,165 @@
+	var/max_size = 100
+	var/last_event = 0
+
+/obj/item/weapon/material/shard/supermatter/New(var/loc, var/level = 1, var/set_size = 0)
```
`/Initialize(var/maploading, var/level = 1, var/set_size = 0)`

PsiOmegaDelta commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,165 @@
+	else
+		size = set_size
+
+	name = "level [smlevel] " + initial(name)
```
Send weird to have the level be part of the name.

PsiOmegaDelta commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,165 @@
+		size = set_size
+
+	name = "level [smlevel] " + initial(name)
+	START_PROCESSING(SSprocessing, src)
```
We have SSObj for processing objects.

PsiOmegaDelta commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,165 @@
+
+	name = "level [smlevel] " + initial(name)
+	START_PROCESSING(SSprocessing, src)
+	update_icon()
```
Not calling `. = ..()`

PsiOmegaDelta commented on code/game/objects/smpulse.dm:
```diff
@@ -0,0 +1,53 @@
+proc/supermatter_delamination(turf/epicenter, devastation_range = 9, smlevel = 1, adminlog = 1, z_transfer = UP|DOWN, shaped)
+	var/multi_z_scalar = 0.35
+	src = null
+	spawn(0)
```
Shouldn't spawn.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,175 @@
+
+/obj/machinery/phoron_desublimer/resonant_chamber/Destroy()
+	for(var/obj/item/sm_piece in sm_pieces)
+		sm_piece.forceMove(loc)
```
`sm_piece.dropInto(loc)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,175 @@
+	for(var/obj/item/sm_piece in sm_pieces)
+		sm_piece.forceMove(loc)
+		sm_piece = null
+	QDEL_NULL_LIST(sm_pieces)
```
This will delete all the shards you just dropped to the ground.

CakeQ replied:
intentional

PsiOmegaDelta replied:
Then why even bother dropping them unto the ground?

PsiOmegaDelta reviewed (changes requested):
There are a heplap of other issues but it's a pain to review on the phone.

CakeQ commented:
@afterthought2 conflicts resolved

halworsen commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,170 @@
+	var/last_event = 0
+
+/obj/item/weapon/material/shard/supermatter/Initialize(var/maploading, var/level = 1, var/set_size = 0)
+	if(level > MAX_SUPERMATTER_LEVEL)
```
`level = Clamp(level, MIN_SUPERMATTER_LEVEL, MAX_SUPERMATTER_LEVEL)`

halworsen commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,170 @@
+	var/datum/sm_control/sm_tier = GLOB.supermatter_tiers[smlevel]
+	color = sm_tier.color
+	light_color = color
+	set_light(0.5, 0.1, light_outer_range, 2, light_color )
```
Whitespace before closing parenthesis

halworsen commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,170 @@
+		icon_state = "supermattersmall"
+		pixel_x = rand(-12, 12)
+		pixel_y = rand(-12, 12)
+	else if( src.size <= 67 )
```
Inconsistent whitespace

halworsen commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,170 @@
+	light_color = color
+	set_light(0.5, 0.1, light_outer_range, 2, light_color )
+
+	if(size <= 34 )
```
Inconsistent whitespace

halworsen commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,170 @@
+		pixel_y = rand(-5, 5)
+
+/obj/item/weapon/material/shard/supermatter/attackby(obj/item/weapon/W as obj, mob/user as mob)
+	if( istype( W, /obj/item/weapon ))
```
More inconsistent whitespace

halworsen commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,170 @@
+
+/obj/item/weapon/material/shard/supermatter/attack_hand(var/mob/user)
+	var/mob/living/carbon/human/H = user
+	if(H.gloves && istype(H.gloves, /obj/item/clothing/gloves/sm_proof))
```
Check for flag instead of type.

halworsen commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,170 @@
+	if(world.time > last_event+20)
+		if(prob(50))
+			SSradiation.radiate(src, 4 * smlevel * (size / max_size))
+		last_event = world.time
```
Nitpicky as hell, but `next_event = world.time + 20` and a check for `world.time > next_event` saves on arithmetic operations.

halworsen commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,170 @@
+	else if( size > 10 )
+		src.visible_message( "The supermatter shard shatters into smaller fragments!" )
+		for( size, size > 10, size -= 10 )
+			new /obj/item/weapon/material/shard/supermatter(src.loc, smlevel)
```
Should probably set the size of the new shards. Since default shard size is 1 and you add between 0 and 10 size on init when size isn't set you can end up with shards whose sizes add up to more than the size of the original shard.

halworsen commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,170 @@
+
+	playsound(src, "shatter", 70, 1)
+	playsound(src, "sound/effects/EMPulse.ogg", 70, 1)
+	qdel(src)
```
`qdel_self()`

PsiOmegaDelta replied:
`qdel(src)` is appropriate here.
`qdel_self()` is for use when listening to events.

halworsen commented on code/game/objects/smpulse.dm:
```diff
@@ -0,0 +1,52 @@
+	epicenter = get_turf(epicenter)
+	if(!epicenter) return
+	if(z_transfer)
+		var/adj_dev   = max(0, (multi_z_scalar * devastation_range) - (shaped ? 2 : 0) )
```
Inconsistent whitespace

halworsen commented on code/game/objects/smpulse.dm:
```diff
@@ -0,0 +1,52 @@
+		if(istype(T, /turf/space) || istype(T, /turf/simulated/open))
+			continue
+
+		var/dist = sqrt((T.x - x0)**2 + (T.y - y0)**2)
```
Just a thought, but maybe this could be approximated for a performance boost?

halworsen commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,175 @@
+			catalyst = B
+			to_chat(user, "You put [B] into the machine.")
+		else
+			to_chat(user, "<span class='notice'>There is alreaddy [catalyst] in the catalyst slot!</span>")
```
already*

halworsen replied:
already*

halworsen commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,175 @@
+			to_chat(user, "<span class='notice'>The machine is full!</span>")
+	else if(istype(B, /obj/item/weapon/metaphoron))
+		if(!catalyst)
+			user.drop_item()
```
```
if(!user.drop_item())
    return
```

afterthought2 replied:
This is incorrect; use unEquip.

halworsen commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,175 @@
+				to_chat(user, "The machine lacks a catalyst and cannot operate.")
+			else if(!active)
+				src.combine_SM()
+			return
```
All these returns are unnecessary

halworsen replied:
Unnecessary returns

halworsen commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,175 @@
+		if("Eject Catalyst")
+			eject_catalyst(user)
+			return
+		if("Cancel")
```
This whole block is unnecessary

halworsen replied:
This block is unnecessary

halworsen commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,175 @@
+			to_chat(user, "<span class='notice'>The machine is full!</span>")
+	else if(istype(B, /obj/item/weapon/metaphoron))
+		if(!catalyst)
+			user.drop_item()
```
```
if(!user.drop_item())
    return
```

afterthought2 replied:
This is incorrect; use unEquip.

halworsen commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,175 @@
+		var/obj/item/weapon/metaphoron/shard = new(src, 1, total_size)
+		sm_pieces += shard
+
+	qdel(catalyst)
```
`QDEL_NULL(catalyst)`

halworsen commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,175 @@
+	if(!catalyst)
+		return
+
+	catalyst.loc = get_turf(src)
```
`catalyst.forceMove(get_turf(src))`

halworsen commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,143 @@
+	ui_interact(user)
+
+/obj/machinery/computer/phoron_desublimer_control/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
+	if(stat & (BROKEN|NOPOWER)) return
```
This is already checked for in `attack_hand`. Remove one of them.

halworsen commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,143 @@
+	else if(href_list["set_neutron_flow"])
+		var/diff = text2num(href_list["set_neutron_flow"])
+		if(diff)
+			furnace.neutron_flow = diff
```
`furnace.neutron_flow = Clamp(diff, furnace.min_neutron_flow, furnace.max_neutron_flow)`

halworsen commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,189 @@
+
+/obj/machinery/phoron_desublimer/furnace/proc/modify_flow(var/change)
+	neutron_flow += change
+	if( neutron_flow > max_neutron_flow )
```
`neutron_flow = Clamp(neutron_flow, 0, max_neutron_flow)`

halworsen commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,189 @@
+	if( !shard )
+		return 0
+
+	qdel(shard)
```
`QDEL_NULL(shard)`

halworsen commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,150 @@
+	return ..()
+
+/obj/machinery/phoron_desublimer/vessel/proc/filled()
+	if(air_contents.total_moles > 1)
```
`return (air_contents.total_moles > 1)`

halworsen commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,150 @@
+
+	if( !loaded_shard ) //If our shard turned into something else, aka full crystal
+		explosion(get_turf(src), 0, 0, 5, 10, 1)
+		qdel(src)
```
`qdel_self()`

halworsen commented on code/modules/supermatter/supermatter.dm:
```diff
@@ -535,6 +579,40 @@
 	log_and_message_admins("WARN: Explosion near the Supermatter! New EER: [power].")
 
+/obj/machinery/power/supermatter/proc/set_sm_level(var/inputlevel = 1)
+	if(inputlevel > MAX_SUPERMATTER_LEVEL)
```
`inputlevel = Clamp(inputlevel, MIN_SUPERMATTER_LEVEL, MAX_SUPERMATTER_LEVEL)`

halworsen commented on code/modules/supermatter/supermatter.dm:
```diff
@@ -535,6 +579,40 @@
+	src.color = smtier.color
+	src.light_color = smtier.color
+
+	name = "level [smlevel] " + initial(name)
```
This makes the science skillcheck redundant

halworsen commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,153 @@
+	light_outer_range = 6
+
+/obj/structure/metaphoron/New(var/loc, var/level = 1)
+	if(level > MAX_SUPERMATTER_LEVEL)
```
`level = Clamp(level, MIN_SUPERMATTER_LEVEL, MAX_SUPERMATTER_LEVEL)`

halworsen commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,153 @@
+/obj/structure/metaphoron
+	name = "supermatter crystal"
+	desc = "A jutting spike of glowing supermatter. Forged from a radioactive baptism, this crystal seems a lot more stable than its' parent counterpart."
```
its*

halworsen commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,153 @@
+	if(world.time > last_event+20)
+		if(prob(50))
+			SSradiation.radiate(src, 2 * smlevel)
+		last_event = world.time
```
Same as with supermatter shard processing.

halworsen commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,153 @@
+				new /obj/item/weapon/metaphoron/low(get_turf(src), smlevel)
+		else
+			visible_message("<span class='alert'><B>\The [src] shatters to dust!</B></span>")
+	qdel(src)
```
`qdel_self()`

halworsen commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,153 @@
+
+/obj/structure/metaphoron/wall/New()
+	. = ..()
+	icon_state = "wall_crystal"
```
Use `update_icon`

halworsen commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,153 @@
+/obj/structure/metaphoron/ex_act(severity)
+	switch(severity)
+		if(1.0)
+			qdel(src)
```
`qdel_self()`

halworsen commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,153 @@
+			return
+		if(2.0)
+			if(prob(50))
+				qdel(src)
```
`qdel_self()`

afterthought2 replied:
`qdel(src)` is preferred, actually; `qdel_self` is a helper for callbacks, observers, and such.

halworsen commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,153 @@
+	var/last_event = 0
+
+/obj/item/weapon/metaphoron/New(var/loc, var/level = 1)
+	if(level > MAX_SUPERMATTER_LEVEL)
```
`level = Clamp(level, MIN_SUPERMATTER_LEVEL, MAX_SUPERMATTER_LEVEL)`

halworsen commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,153 @@
+	if(world.time > last_event+20)
+		if(prob(50))
+			SSradiation.radiate(src, 15 * smlevel)
+		last_event = world.time
```
Same as with supermatter shard processing.

halworsen commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,175 @@
+				playsound(loc, 'sound/effects/metalhit.ogg', 75, 1)
+			health -= B.force
+			if(health <= 0)
+				qdel(src)
```
`qdel_self()`

halworsen commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,175 @@
+			playsound(loc, 'sound/effects/metalhit.ogg', 75, 1)
+		health -= B.force
+		if(health <= 0)
+			qdel(src)
```
`qdel_self()`

halworsen commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,175 @@
+/obj/machinery/phoron_desublimer/ex_act(severity)
+	switch(severity)
+		if(1.0)
+			qdel(src)
```
`qdel_self()`

halworsen commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,175 @@
+			return
+		if(2.0)
+			if(prob(50))
+				qdel(src)
```
`qdel_self()`

halworsen reviewed (commented):
I gave up on the whitespace comments. Pls fix, it's kinda ugly.

afterthought2 commented:
I'm hitting this with "Last Chance." It's been sitting here ignored for a while, and that isn't very kind to the author. I would be willing to review the code further, but only if there's another dev that expresses willingness to approve/merge it down the road.

PsiOmegaDelta commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,167 @@
+
+/obj/item/weapon/material/shard/supermatter/Destroy()
+	STOP_PROCESSING(SSobj, src)
+	..()
```
Not returning the result of ..()

PsiOmegaDelta commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,167 @@
+	if(istype(W, /obj/item/weapon))
+		if(W.force >= 5)
+			src.shatter()
+	..()
```
Don't think it's a good idea to call `..()` in `attackby()` for objects that may have just deleted themselves.

PsiOmegaDelta commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,167 @@
+/obj/item/weapon/material/shard/supermatter/attackby(obj/item/weapon/W as obj, mob/user as mob)
+	if(istype(W, /obj/item/weapon))
+		if(W.force >= 5)
+			src.shatter()
```
Noting that the use of `src` is superfluous more often than not.

PsiOmegaDelta commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,167 @@
+							"<span class='alert'>You aren't sure that's a good idea.</span>",
+							"<span class='alert'>You aren't in the mood to get vaporized today.</span>",
+				 			"<span class='alert'>You really don't feel like frying your hand off.</span>",
+							"<span class='alert'>You assume that's a bad idea.</span>"))
```
Why not pick the message and then do `to_chat(user, "<span class='alert'>[message]</span>"`?

PsiOmegaDelta replied:
Typically `danger` is the standard span class for messages like these.

PsiOmegaDelta commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,167 @@
+	STOP_PROCESSING(SSobj, src)
+	..()
+
+/obj/item/weapon/material/shard/supermatter/examine(mob/user)
```
Not calling `..()`

PsiOmegaDelta commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,167 @@
+		if(L)
+			fuse_limb(user, L)
+			shatter()
+	..()
```
Again, probably not wise to call `..()` for an object that may have just deleted itself.

PsiOmegaDelta commented on code/game/objects/smpulse.dm:
```diff
@@ -0,0 +1,64 @@
+proc/supermatter_delamination(turf/epicenter, devastation_range = 9, smlevel = 1, adminlog = 1, z_transfer = UP|DOWN, shaped)
```
Missing initial `/`

PsiOmegaDelta commented on code/game/objects/smpulse.dm:
```diff
@@ -0,0 +1,64 @@
+proc/supermatter_delamination(turf/epicenter, devastation_range = 9, smlevel = 1, adminlog = 1, z_transfer = UP|DOWN, shaped)
+	var/multi_z_scalar = 0.35
+	src = null
```
..why?

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,137 @@
+
+/obj/machinery/computer/phoron_desublimer_control/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
+	if(stat & (BROKEN|NOPOWER)) return
+	if(user.stat || user.restrained()) return
```
`Topic()` handles general checks like these.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,137 @@
+	ui_interact(user)
+
+/obj/machinery/computer/phoron_desublimer_control/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
+	if(stat & (BROKEN|NOPOWER)) return
```
`CanUseTopic()` handles special checks like these.
This check is already in `/obj/machinery/CanUseTopic()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,137 @@
+		frequency_presets[preset] = furnace.neutron_flow
+
+/obj/machinery/computer/phoron_desublimer_control/attack_ai(mob/user)
+	if(stat & (BROKEN|NOPOWER))
```
Already handled in `CanUseTopic()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,137 @@
+			vessel = PD
+			vessel.com = src
+		if(istype(PD, /obj/machinery/phoron_desublimer/furnace))
+			furnace = PD
```
Not handling the furnace being deleted.

afterthought2 replied:
See my comment below. You are handling it, but it needs improvement.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,137 @@
+/obj/machinery/computer/phoron_desublimer_control/proc/find_parts()
+	for(var/obj/machinery/phoron_desublimer/PD in get_area(src))
+		if(istype(PD, /obj/machinery/phoron_desublimer/vessel))
+			vessel = PD
```
Not handling the vessel being deleted.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,171 @@
+	if(!catalyst)
+		return
+
+	catalyst.forceMove(get_turf(src))
```
`catalyst.dropInto(loc)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,171 @@
+
+	var/max_pieces = 8
+	var/list/sm_pieces = list()
+	var/obj/item/weapon/metaphoron/catalyst = null
```
Not handling a set `catalyst` in `/Destroy()`

PsiOmegaDelta commented on code/game/objects/smpulse.dm:
```diff
@@ -0,0 +1,64 @@
+				new /obj/structure/metaphoron(T, smlevel)
+				spawned = TRUE
+	var/took = (world.timeofday-start)/10
+	if(Debug2) world.log << "## DEBUG: Explosion([x0],[y0],[z0])(d[devastation_range],level [smlevel]): Took [took] seconds."
```
We have appropriate debug procs for messages such as these, such as `log_debug()` for example.

PsiOmegaDelta commented on code/game/objects/smpulse.dm:
```diff
@@ -0,0 +1,64 @@
+	far_dist += devastation_range * 20
+	var/frequency = get_rand_frequency()
+	for(var/mob/M in GLOB.player_list)
+		if(M.z == epicenter.z)
```
This check will be unable to handle mobs inside other objects.

CakeQ replied:
The classic Indiana Jones hiding in a fridge strategy. I think I'm alright with this.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,171 @@
+		else
+			user.visible_message("[user] starts unfastening the bolts on \the [src].", "You start unfastening the bolts on \the [src].")
+
+		if(do_after(user, 20))
```
`do_after(user, 20, src)` and the progress bar appears over the appropriate object and you won't need the deleted check below.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,171 @@
+			else
+				playsound(loc, 'sound/effects/metalhit.ogg', 75, 1)
+			health -= B.force
+			if(health <= 0)
```
Copy-pasty health checks all over this file.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,171 @@
+	anchored = 1
+
+	var/max_pieces = 8
+	var/list/sm_pieces = list()
```
`list()`s should be assigned in `/Initialize()`, not the type definition.
The `LAZY_` macros are preferred over manually assigning a list at all.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,171 @@
+
+	if(istype(B, /obj/item/weapon/material/shard/supermatter))
+		if(sm_pieces.len < max_pieces)
+			if(!user.drop_item())
```
``` 
if(!user.unequip_item(src))
    return
```
Can then remove the `forceMove()` line below as well.

afterthought2 replied:
This isn't resolved. `drop_item` is incorrect; you should do what Psi said or `if(!unEquip(B, src)) return`.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,171 @@
+
+/obj/machinery/phoron_desublimer/ex_act(severity)
+	switch(severity)
+		if(1.0)
```
The `.0` is from an ancient decompilation of  SS13 code.
Do clean up and use pure integers.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,171 @@
+			else
+				src.visible_message("\icon[src] <b>[src]</b> beeps, \"Machine is too full to load \the [AM].\"")
+
+	..()
```
Should there really be potentially other bump handling when the target has been moved into the chamber?

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,171 @@
+				qdel(src)
+			return
+
+/obj/machinery/phoron_desublimer/resonant_chamber/Bumped(atom/AM as mob|obj)
```
`/obj/machinery/phoron_desublimer/resonant_chamber/Bumped(var/atom/movable/AM)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,171 @@
+	QDEL_NULL_LIST(sm_pieces)
+	. = ..()
+
+/obj/machinery/phoron_desublimer/resonant_chamber/attackby(var/obj/item/weapon/B as obj, var/mob/user as mob)
```
Drop `as obj` and `as mob`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,171 @@
+	active = 0
+
+/obj/machinery/phoron_desublimer/resonant_chamber/proc/eject(mob/user as mob)
+	var/obj/sm_piece = input(user,	"What would you like to eject?", \
```
Not checking that the user is able to complete the action after the `input()` is conducted.
`CanPhysicallyInteract(user)` is a handy proc, that's better than the copy-pasting of adjacency and incapacitation checks (and also covers abilities such as telekinesis).

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,137 @@
+		return
+	var/preset = null
+	preset = input(user, "Which preset would you like set?", "Select Preset", preset) in frequency_presets
+	if(!user.incapacitated() && user.Adjacent(loc))
```
`if(CanPhysicallyInteract(user))`
Repeat everywhere.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,137 @@
+		return 0
+	return 1
+
+/obj/machinery/computer/phoron_desublimer_control/proc/set_preset(var/mob/user as mob)
```
`var/mob/user`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,137 @@
+		ui.open()
+		ui.set_auto_update(1)
+
+/obj/machinery/computer/phoron_desublimer_control/Topic(href, href_list)
```
Should use `/obj/machinery/computer/phoron_desublimer_control/OnTopic(user, href_list)`
Then you no longer have to check `..()` but must instead return valid `TOPIC_` defines.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+								MATERIAL_OSMIUM = 1,
+								MATERIAL_SUPERMATTER = 1.0) // modifier for output amount
+
+	var/list/mat_peak = list()
```
Unnecessary assignment of list as it's overridden in `/Initialize()` anyway.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+	var/list/mat_peak = list()
+
+	var/list/obj/item/stack/material/mat_obj = list(MATERIAL_DIAMOND = /obj/item/stack/material/diamond,
+													MATERIAL_STEEL = /obj/item/stack/material/steel,
```
Far as I can tell, looking at the diff, the whitespace offset here is ridiculous.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+
+/obj/machinery/phoron_desublimer/furnace/Destroy()
+	if(shard)
+		shard.forceMove(loc)
```
`shard.dropInto(loc)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+		src.visible_message("\icon[src] <b>[src]</b> buzzes, \"Needs a supermatter shard to transmutate.\"")
+		return
+
+	active = 1
```
The `produce()` proc itself should check if `active` is set or not before proceeding.
Same with any and all other procs with activity flags like this.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+
+				if(amount > 0) // Will only do anything if any amount was actually created
+					var/obj/item/stack/material/T = mat_obj[cur_mat]
+					var/obj/item/stack/material/I = new T
```
`var/obj/item/stack/material/I = new T(loc, amount)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+		return 0
+
+	QDEL_NULL(shard)
+	shard = null
```
..what do you think `QDEL_NULL()` do?
Superfluous null-setting.

CakeQ replied:
was originally just a qdel, I glossed over it

PsiOmegaDelta commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,183 @@
+		to_chat(user, "<span class='notice'>This machine only accepts supermatter shards</span>")
+
+	update_icon()
+	return
```
Superfluous return.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_machine.dm:
```diff
@@ -0,0 +1,40 @@
+		ready = 0
+		return
+
+	ready = 1
```
```
ready = !(stat & (BROKEN|NOPOWER))
return ready
```
Then you can remove the meaningless `/obj/machinery/phoron_desublimer/furnace/report_ready()` override.

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,147 @@
+
+/obj/machinery/phoron_desublimer/vessel/Destroy()
+	com = null
+	qdel(loaded_tank)
```
`QDEL_NULL()`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,147 @@
+	QDEL_NULL(air_contents)
+	. = ..()
+
+/obj/machinery/phoron_desublimer/vessel/on_update_icon(var/state = 1)
```
..why does this proc depend on a passed value instead of the object state?

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,147 @@
+	com = null
+	qdel(loaded_tank)
+	if(loaded_shard)
+		loaded_shard.forceMove(loc)
```
`loaded_shard.dropInto(loc)`

PsiOmegaDelta commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,147 @@
+		return
+
+	src.visible_message("\icon[src] <b>[src]</b> beeps, \"Feeding crystal.. Do not remove.\"")
+	active = 1
```
Again, not checking its own activity flag before proceeding.

PsiOmegaDelta commented on code/modules/supermatter/supermatter.dm:
```diff
@@ -104,9 +107,19 @@
+	var/smlevel = 1
+	var/psionic_power = 20 //Severity of hallucinations when looking at the core.
+
+/obj/machinery/power/supermatter/New(var/loc, var/level = 1)
```
`/obj/machinery/power/supermatter/Initialize(var/maploading, var/level = 1)`

PsiOmegaDelta commented on code/modules/supermatter/supermatter.dm:
```diff
@@ -104,9 +107,19 @@
+	var/psionic_power = 20 //Severity of hallucinations when looking at the core.
+
+/obj/machinery/power/supermatter/New(var/loc, var/level = 1)
 	..()
```
`. = ..()`

PsiOmegaDelta commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,147 @@
+	light_color = COLOR_SM_DEFAULT
+	light_outer_range = 6
+
+/obj/structure/metaphoron/New(var/loc, var/level = 1)
```
`/obj/structure/metaphoron/Initialize(var/maploading, var/level = 1)`

PsiOmegaDelta replied:
Not doing `. = ..()`

PsiOmegaDelta commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,147 @@
+
+/obj/structure/metaphoron/Destroy()
+	STOP_PROCESSING(SSprocessing, src)
+	..()
```
`. = ..()`

PsiOmegaDelta commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,147 @@
+	STOP_PROCESSING(SSprocessing, src)
+	..()
+
+/obj/structure/metaphoron/examine(mob/user)
```
Not calling `..()`

PsiOmegaDelta commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,147 @@
+	START_PROCESSING(SSprocessing, src)
+	update_icon()
+
+/obj/structure/metaphoron/random/highlevel/New()
```
`/Initialize()`

PsiOmegaDelta commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,147 @@
+	update_icon()
+
+/obj/structure/metaphoron/random/highlevel/New()
+	..()
```
`. = ..()`

PsiOmegaDelta commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,147 @@
+
+/obj/item/weapon/metaphoron/Destroy()
+	STOP_PROCESSING(SSprocessing, src)
+	..()
```
`. = ..()`

PsiOmegaDelta commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,147 @@
+/obj/structure/metaphoron
```
Metaphoron feels like a rather copy-pasty version `/obj/item/weapon/material/shard/supermatter`

CakeQ replied:
They have different purposes

afterthought2 replied:
But they share virtually all of their code. Just make them subtypes of a common type, unless there's something you are gaining from this being a structure.

PsiOmegaDelta commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,147 @@
+	icon_state = "wall_crystal"
+	density = 0
+
+/obj/structure/metaphoron/wall/New()
```
`/Initialize()`

PsiOmegaDelta commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,147 @@
+				return
+
+
+/obj/structure/metaphoron/random/New()
```
``` 
.../Initialize()
    . = ..()
```

PsiOmegaDelta commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,147 @@
+	var/smlevel = 1
+	var/next_event = 0
+
+/obj/item/weapon/metaphoron/New(var/loc, var/level = 1)
```
```
/Initialize(var/maploading, var/level = 1)
    . = ..()
```

afterthought2 commented:
@PsiOmegaDelta  appears to have shown interest, so I'm removing the last chance tag.

babydoll commented:
Merge conflicts and review to address.

CakeQ commented:
@PsiOmegaDelta comments addressed, hopefully ready for approval

afterthought2 commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,171 @@
+	level = Clamp(level, MIN_SUPERMATTER_LEVEL, MAX_SUPERMATTER_LEVEL)
+
+	smlevel = level
+	light_color = rgb(255, 0, (255-(255*(smlevel/MAX_SUPERMATTER_LEVEL))))
```
Need to round, I believe.

Upon further reading, am confused: you are about to call `update_icon`, which appears to reset the color based on a global datum anyway.

afterthought2 commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,171 @@
+	. = ..()
+
+/obj/item/weapon/material/shard/supermatter/examine(mob/user)
+	..()
```
Has a meaningful return, that I believe needs to be checked.

afterthought2 commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,171 @@
+		pixel_x = rand(-5, 5)
+		pixel_y = rand(-5, 5)
+
+/obj/item/weapon/material/shard/supermatter/attackby(obj/item/weapon/W as obj, mob/user as mob)
```
Remove `as obj`, `as mob`.

afterthought2 commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,171 @@
+							"You assume that's a bad idea.")
+		to_chat(user, "<span class='danger'>[message]</span>")
+		return
+	..()
```
I don't get it; where does the bad stuff happen? this seems to have the same effect as touching it while protected.

afterthought2 commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,171 @@
+		if(prob(50))
+			SSradiation.radiate(src, 4 * smlevel * (size / max_size))
+		next_event = world.time + 20
+	..()
```
I doubt that this parent call is a good idea.

afterthought2 commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,171 @@
+
+/obj/item/weapon/material/shard/supermatter/shatter()
+	if(size > 100)
+		src.visible_message("The supermatter shard grows into a full-sized supermatter crystal!")
```
Get rid of unneeded `src`; also below.

afterthought2 commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,171 @@
+	update_icon()
+
+/obj/item/weapon/material/shard/supermatter/shatter()
+	if(size > 100)
```
Shouldn't this be `max_size`?

afterthought2 commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,171 @@
+		for(size, size > 10, size -= 10)
+			new /obj/item/weapon/material/shard/supermatter(src.loc, smlevel)
+	else
+		src.visible_message("The supermatter shard shatters into dust!")
```
So it seems like if you are in the 10-19 range, you'll shatter into...one shard? Should maybe not qdel self, but rather reset self's size to the remainder, maybe with some further random loss.

afterthought2 commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,171 @@
+
+/obj/item/weapon/material/shard/supermatter/equipped(var/mob/user)
+	var/mob/living/carbon/human/H = user
+	var/equipped_slot = user.get_inventory_slot(src)
```
Equipped has a second argument it's called with, which is the slot into which it was equipped.

afterthought2 commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,171 @@
+			fuse_limb(user, L)
+			shatter()
+	else
+		..()
```
I feel it's safer to just always call `..()` (but first, before any of this stuff). It does a bunch of icon and UI operations, which are presumably not needed here, but it might be that you're putting it into an unexpected state by not calling it.

afterthought2 commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,171 @@
+					return
+			fuse_limb(L, O)
+		else
+			L.dust()
```
Seems disproportionately severe for borgs.

afterthought2 commented on code/game/objects/items/weapons/material/shards.dm:
```diff
@@ -100,3 +100,171 @@
+			var/mob/living/carbon/human/H = L
+			var/obj/item/organ/external/O = pick(H.organs)
+
+			for(var/obj/item/clothing/C in list(H.head, H.wear_mask, H.wear_suit, H.w_uniform, H.gloves, H.shoes))
```
Try `get_covering_equipped_item`. May also be useful for the equip code just above.

In fact, maybe move this whole check to `fuse_limb`.

afterthought2 commented on code/game/objects/smpulse.dm:
```diff
@@ -0,0 +1,64 @@
+/proc/supermatter_delamination(turf/epicenter, devastation_range = 9, smlevel = 1, adminlog = 1, z_transfer = UP|DOWN, shaped)
```
This is a straight copypaste of `explosion`. If there are some minor differences (I can't tell) then move this thing to a datum that hosts the explosion. Can then also have it `Process`, and check tick, reducing lag for everyone.

afterthought2 commented on code/game/objects/smpulse.dm:
```diff
@@ -0,0 +1,64 @@
+	sleep(8)
+	return 1
+
+proc/approx_dist(var/x, var/y)
```
Computing squared Euclidean distance is very fast; so long as you don't do the square root, it may in fact be faster than this. In any case, we can afford it.

afterthought2 commented on code/modules/clothing/gloves/miscellaneous.dm:
```diff
@@ -143,3 +143,11 @@
+	item_state = "sm_proof"
+	permeability_coefficient = 0.9
+	siemens_coefficient = 0.9
+	item_flags = ITEM_FLAG_SMPROOF
```
If these items are going to be found on away maps, you probably want to give this flag to appropriate spacesuits.

afterthought2 commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,173 @@
+				user.visible_message("[user] unfastens the bolts securing \the [src] to the floor.", "You unfasten the bolts securing \the [src] to the floor.")
+				anchored = 0
+		else
+			user.setClickCooldown(DEFAULT_ATTACK_COOLDOWN)
```
I think, despite oldcode, doing this manually in random places is almost never correct.

afterthought2 commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,173 @@
+		else
+			user.setClickCooldown(DEFAULT_ATTACK_COOLDOWN)
+			if(B.damtype == BRUTE || B.damtype == BURN)
+				user.do_attack_animation(src)
```
Wait, what? If you break the do_after when attacking it with a wrench, that deals damage, and that's the only thing that does? I don't get it.

I suggest just letting base /obj code handle wrenching/unwrenching, as it seems more sane anyway.

afterthought2 commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,173 @@
+			to_chat(user, "<span class='notice'>The machine is full!</span>")
+	else if(istype(B, /obj/item/weapon/metaphoron))
+		if(!catalyst)
+			if(!user.drop_item())
```
See above. `drop_item` is incorrect.

afterthought2 commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,173 @@
+/obj/machinery/phoron_desublimer/resonant_chamber/Bumped(var/atom/movable/AM)
+	if(anchored)
+		if(istype(AM, /obj/machinery/power/supermatter))
+			if(sm_pieces.len < max_pieces)
```
Not the first time I'm seeing this stuff. Pull it out into a proc: `try_add_supermatter`, returns true if succeeds for further handling.

afterthought2 commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,173 @@
+	else
+		..()
+
+/obj/machinery/phoron_desublimer/resonant_chamber/attack_hand(mob/user as mob)
```
Kill the `as_mob`.

afterthought2 commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,173 @@
+		return
+	var/choice = input(user,"What would you like to do?", \
+							"Resonant Chamber", \
+							"Cancel") in list("Combine", "Eject", "Eject Catalyst", "Cancel")
```
I'll be honest, I don't know the syntax well enough to know what this defaults to. Do `as null|anything` to make it easier to read.

afterthought2 commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,173 @@
+	var/choice = input(user,"What would you like to do?", \
+							"Resonant Chamber", \
+							"Cancel") in list("Combine", "Eject", "Eject Catalyst", "Cancel")
+	if(user.incapacitated() || !user.Adjacent(loc) || !anchored || active)
```
The before and after checks are not the same for some reason. Use `CanPhysicallyInteract` or similar.

afterthought2 commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,173 @@
+	anchored = 1
+
+	var/max_pieces = 8
+	var/list/sm_pieces
```
Is there any point in having this, versus just iterating through contents? You're going to have a GC issue with this, as when one of them is deleted, it will fail because this is holding on to a ref, and it just seems unnecessary.

afterthought2 commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,173 @@
+		sm_pieces += shard
+
+	QDEL_NULL(catalyst)
+	catalyst = null
```
you just nulled it.

afterthought2 commented on code/modules/phoron_desub/desub_chamber.dm:
```diff
@@ -0,0 +1,173 @@
+	playsound(loc, 'sound/machines/ding.ogg', 50, 1, -1)
+	active = 0
+
+/obj/machinery/phoron_desublimer/resonant_chamber/proc/eject(mob/user as mob)
```
Kill all `as mob` everywhere.

afterthought2 commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,129 @@
+		frequency_presets[preset] = furnace.neutron_flow
+
+/obj/machinery/computer/phoron_desublimer_control/attack_ai(mob/user)
+	interact(user)
```
Shouldn't this be `ui_interact`?

afterthought2 commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,129 @@
+	var/preset = null
+	preset = input(user, "Which preset would you like set?", "Select Preset", preset) in frequency_presets
+	if(CanPhysicallyInteract(user))
+		frequency_presets[preset] = furnace.neutron_flow
```
Is this really what you mean? How do you add stuff to the list?

afterthought2 commented on code/modules/phoron_desub/desub_control.dm:
```diff
@@ -0,0 +1,129 @@
+	else if(href_list["furnace_eject_shard"])
+		furnace.eject_shard()
+	else if(href_list["neutron_adj"])
+		var/diff = text2num(href_list["neutron_adj"])
```
You probably want to `sanitize_integer`.

afterthought2 commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,180 @@
+	for(var/obj/machinery/computer/phoron_desublimer_control/DC in get_area(src))
+		if(DC.furnace)
+			DC.furnace = null
+			DC.find_parts()
```
It's just going to find us again; we aren't in nullspace yet. Anyway, this is terrible. Suggested way to handle this: use `GLOB.destroyed_event.register(furnace, src, .proc/remove_furnace)`.

afterthought2 commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,180 @@
+				//amount += amount*size_modifier
+
+				var/k = (2 * M_PI) / (max_distance * 4)
+				var/sin_exp = (k * M_PI * distance) + (M_PI / 2)
```
The squared pi seems like a mistake.

afterthought2 commented on code/modules/phoron_desub/desub_furnace.dm:
```diff
@@ -0,0 +1,180 @@
+	return 1
+
+/obj/machinery/phoron_desublimer/furnace/attackby(var/obj/item/weapon/B as obj, var/mob/user as mob)
+	if(isrobot(user))
```
Why? Let them do their thing.

afterthought2 commented on code/modules/phoron_desub/desub_machine.dm:
```diff
@@ -0,0 +1,41 @@
+/*
+///////////// PHORON DESUBLIMER ////////////////
```
Move this stuff to a codex entry.

afterthought2 commented on code/modules/phoron_desub/desub_machine.dm:
```diff
@@ -0,0 +1,41 @@
+	icon = 'icons/obj/phoron_desublimation.dmi'
+	density = 1
+	anchored = 1
+	use_power = 1
```
use an appropriate define.

afterthought2 commented on code/modules/phoron_desub/desub_machine.dm:
```diff
@@ -0,0 +1,41 @@
+	var/active = 0
+
+/obj/machinery/phoron_desublimer/Process()
+	if( stat & ( BROKEN|NOPOWER ))
```
Have appropriate procs set ready; no need to constantly process for no reason.

afterthought2 commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,138 @@
+		loaded_shard.dropInto(loc)
+		loaded_shard = null
+	QDEL_NULL(air_contents)
+	for(var/obj/machinery/computer/phoron_desublimer_control/DC in get_area(src))
```
See my comment about the furnace.

afterthought2 commented on code/modules/phoron_desub/desub_vessel.dm:
```diff
@@ -0,0 +1,138 @@
+	sleep(22)
+	update_icon()
+
+	if(!loaded_shard) //If our shard turned into something else, aka full crystal
```
If you legitimately expect the shard to sometimes qdel, you need to subscribe to its destroyed_event, and have it call a proc on you that clears it from the loaded_shard var (in addition to blowing you up and whatever).

afterthought2 commented on code/modules/supermatter/supermatter.dm:
```diff
@@ -238,15 +264,20 @@
+	// Effect 5: Crystal wave
+
+	spawn(0)
+		supermatter_delamination(TS, delamination_size, smlevel)
```
use `INVOKE_ASYNC`.

afterthought2 commented on code/modules/supermatter/supermatter.dm:
```diff
@@ -238,15 +264,20 @@
+
 /obj/machinery/power/supermatter/examine(mob/user)
 	. = ..()
+	if(user.skill_check(SKILL_SCIENCE, SKILL_ADEPT))
```
Eh, keep it engines (though can tone down the level if you want). It's like the one thing they get.

afterthought2 commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,151 @@
+	icon_state = "floor_crystal"
+	density = 1
+	w_class = ITEM_SIZE_LARGE
+	anchored = 1.0
```
`1`, please.

afterthought2 commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,151 @@
+	. = ..()
+
+/obj/structure/metaphoron/examine(mob/user)
+	..()
```
need to return it too, and probably check the return.

afterthought2 commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,151 @@
+	color = sm_tier.color
+	light_color = color
+	set_light(0.5, 0.1, light_outer_range, 2, light_color)
+	START_PROCESSING(SSprocessing, src)
```
objs process on `SSobj`.

afterthought2 commented on code/modules/supermatter/supermatter_debris.dm:
```diff
@@ -0,0 +1,151 @@
+		if(prob(50))
+			SSradiation.radiate(src, 15 * smlevel)
+		next_event = world.time + 20
+	..()
```
Yet another copypasted version? This one can definitely be combined with shards.

afterthought2 reviewed (changes requested):
Seems like a lot of these phoron types should all just be subtypes of a common parent.

afterthought2 commented:
@CakeQ

afterthought2 commented:
Closing as per dev policy. Ask on discord if you'd like it reopened once it's ready to go. I realize we were pretty slow in reviewing this, but that's because none of us were particularly thrilled about the idea to begin with (and because it had a lot of issues).
